### PR TITLE
feat: Add FT weight-only quantization GEMM CUDA kernel support

### PR DESCRIPTION
**Pull request by SidaZh**
# What does this PR do?
Add FT weight-only quantization CUDA kernel support. We extracted weight-only GEMM kernel from [FasterTransformer](https://github.com/NVIDIA/FasterTransformer/tree/main/src/fastertransformer/kernels/cutlass_kernels/fpA_intB_gemm), 
and create a [project](https://github.com/NetEase-FuXi/EETQ.git) to use it easily.
- Features
  * Weight-only quantization is easy to use and no additional training required.
  * Work on every model
  * There is almost no loss of quantization accuracy.
  * Compared to bitsandbytes and gptq, the performance is also better.

- Usage
```bash
$ --quantize eetq
```

- Test
Below is our test on 3090. environment: torch=2.01, cuda=11.8, nvidia driver: 525.78.01
prompt=1024, max_new_tokens=50

benchmark

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by dingjingzhen — CHANGES_REQUESTED** [review]
good job

**Comment by NetEase-FuXi** [comment]
We did benchmark with FT weight-only gemm kernel on single A100
image

**Review by Narsil — COMMENTED** [review]
This PR is very nice, the code is simple, agnostic to TP sharding and indeed much more efficient than BNB.

Imo, we should probably replace BNB (8bits) with this instead.

The only reason why we might not is because bnb can be used with raw transformers (non flash models). Those have poor performance anyway so they are not really our main interest.

I think we could suggest deprecating BNB after this lands.

**Comment by Narsil** [comment]
By the way, we do not really care that much for throughput.  Latency is our key metric. And decode latency even more. (Throughput just happens to come for free when latency is good because LLMs are memory bound)

**Comment by Narsil** [comment]
I can take care of the modifications + rebase if you want btw.

File: launcher/src/main.rs [hunk 1]
```diff
@@ -25,6 +25,7 @@ enum Quantization {
     BitsandbytesNF4,
     BitsandbytesFP4,
     Gptq,
+    WeightOnly,
```
[the remark this answers is not in the document]
Comment (SidaZh) [reply launcher/src/main.rs:28]: Yes, we are totally willing to do that.

**Comment by SidaZh** [comment]
> I can take care of the modifications + rebase if you want btw.

@Narsil 
Thanks very much for your reply. We will improve and modify our code. We are honored that you can directly modify and incorporate the code.

**Comment by NetEase-FuXi** [comment]
> By the way, we do not really care that much for throughput. Latency is our key metric. And decode latency even more. (Throughput just happens to come for free when latency is good because LLMs are memory bound)

We use benchmark to test the lantency of using TGI+EETQ.
image

**Comment by SidaZh** [comment]
> By the way, we do not really care that much for throughput. Latency is our key metric. And decode latency even more. (Throughput just happens to come for free when latency is good because LLMs are memory bound)
We use benchmark to test the latency of using TGI+EETQ.
image

**Comment by Narsil** [comment]
> We will improve and modify our code. We are honored that you can directly modify and incorporate the code.

Something must have been lost in translation. I will do some further PR for the rebase.

**Comment by Narsil** [comment]
https://github.com/huggingface/text-generation-inference/pull/1068 for the rebased branch

**Comment by SidaZh** [comment]
> Okay closing this as the superspeeded PR is merged @SidaZh you are co-author !

Thanks again for your work in merging this PR. I believe it will be helpful for users to use TGI to speed up quantization conveniently.

**Comment by Narsil** [comment]
Indeed it's super useful, I don't know if you've seen but there's now a warning on bitsandbytes recommending to use eetq instead since it's so much faster.

**Comment by OlivierDehaene** [comment]
Very nice PR! Thanks for the hard work!

**Comment by jerryMeng100** [comment]
Hi - this almost seems too good to be true after testing it out for a bit, I haven't seen any performance degradations so far but just wondering - I know bitsandbytes had some work in llm.int8 on outliers and preserving precision for those to retain accuracy, does eetq do the same? If not, does it suffer from the issue outlined in the llm.int8 paper of larger models having worse performance due to outliers? (https://arxiv.org/pdf/2208.07339.pdf) 

Maybe I'm not quite understanding it but is there any performance diff between bitsandbytes 8 bit vs eetq 8 bit, if there is any, or if it just strictly a faster version of bitsandbytes 8 bit. Thanks! @SidaZh

**Comment by SidaZh** [comment]
> Hi - this almost seems too good to be true after testing it out for a bit, I haven't seen any performance degradations so far but just wondering - I know bitsandbytes had some work in llm.int8 on outliers and preserving precision for those to retain accuracy, does eetq do the same? If not, does it suffer from the issue outlined in the llm.int8 paper of larger models having worse performance due to outliers? (https://arxiv.org/pdf/2208.07339.pdf)
> 
> Maybe I'm not quite understanding it but is there any performance diff between bitsandbytes 8 bit vs eetq 8 bit, if there is any, or if it just strictly a faster version of bitsandbytes 8 bit. Thanks! @SidaZh

eetq 8 bit can be regarded as a high-performance cutlass implementation of w8a16(per-channel). From the perspective of the PPL, even without correcting for outliers, the performance of the w8a16 per-channel quantization is already very good.

**Pull request closed without merging**